### PR TITLE
Native assets support for Android Add2app

### PR DESCRIPTION
**Pull request by dcharkes**
Support for FFI calls with @Native external functions through Native assets on Android add to app. This enables bundling native code without any build-system boilerplate code.

For more info see:

*  https://github.com/flutter/flutter/issues/129757

## Implementation details for Android add2app

The `.so` files are bundled with the same mechanism that bundles `libapp.so`.

## Pre-launch Checklist

- [x] I read the [Contributor Guide] and followed the process outlined there for submitting PRs.
- [x] I read the [Tree Hygiene] wiki page, which explains my responsibilities.
- [x] I read and followed the [Flutter Style Guide], including [Features we expect every widget to implement].
- [x] I signed the [CLA].
- [x] I listed at least one issue that this PR fixes in the description above.
- [x] I updated/added relevant documentation (doc comments with `///`).
- [x] I added new tests to check the change I am making, or this PR is [test-exempt].
- [x] All existing and new tests are passing.

[Contributor Guide]: https://github.com/flutter/flutter/wiki/Tree-hygiene#overview
[Tree Hygiene]: https://github.com/flutter/flutter/wiki/Tree-hygiene
[test-exempt]: https://github.com/flutter/flutter/wiki/Tree-hygiene#tests
[Flutter Style Guide]: https://github.com/flutter/flutter/wiki/Style-guide-for-Flutter-repo
[Features we expect every widget to implement]: https://github.com/flutter/flutter/wiki/Style-guide-for-Flutter-repo#features-we-expect-every-widget-to-implement
[CLA]: https://cla.developers.google.com/
[flutter/tests]: https://github.com/flutter/tests
[breaking change policy]: https://github.com/flutter/flutter/wiki/Tree-hygiene#handling-breaking-changes
[Discord]: https://github.com/flutter/flutter/wiki/Chat

### CODE REVIEW

**Comment by dcharkes** [comment]
@reidbaker @bartekpacia friendly ping!

**Review by reidbaker — APPROVED** [review]

**Comment by dcharkes** [comment]
@bartekpacia I'd like your eyes on this as well as you rewrote packages/flutter_tools/gradle/src/main/groovy/flutter.groovy

File: packages/flutter_tools/gradle/src/main/groovy/flutter.groovy [hunk 1]
```diff
@@ -1140,10 +1138,20 @@ class FlutterPlugin implements Plugin<Project> {
...
                         }
                     }
+                    // Copy the native assets created by build.dart and placed in build/native_assets by flutter assemble.
+                    // The `$project.buildDir` is '.android/Flutter/build/' instead of 'build/'.
+                    def buildDir = "${getFlutterSourceDirectory()}/build"
+                    def nativeAssetsDir = "${buildDir}/native_assets/android/jniLibs/lib"
+                    from("${nativeAssetsDir}/${abi}") {
+                        include "*.so"
+                        rename { String filename ->
+                            return "lib/${abi}/${filename}"
```
Comment (bartekpacia) [inline packages/flutter_tools/gradle/src/main/groovy/flutter.groovy:1148]: Small nit:

Tasks in Gradle should not directly access (e.g. copy, move) output artifacts of other tasks. This is what's happening here: the `packJniLibsTask` directly accesses `*.so` files in `build/native_assets/android/jniLibs/lib` (and thus depends on whatever task created that directory).

Instead, if you want to make task A depend on task B, you should use implicit dependencies between tasks and let Gradle figure out the dependency graph for you. In other words, using `dependsOn` is usually a code smell, as seen above in line 1131.

I find that [this blogpost from a former Gradle engineer](https://melix.github.io/blog/2021/10/gradle-quickie-dependson.html) explains this topic well.

This is a nit and not directly related to this PR at all, but I just wanna raise awareness of this issue - it occurs in almost all tasks of Flutter Gradle Plugin. A while back I created https://github.com/flutter/flutter/issues/123934 to track this issue.

**Review by bartekpacia — APPROVED** [review]
I checked only Gradle code and LGTM.

File: packages/flutter_tools/gradle/src/main/groovy/flutter.groovy (hunk 1 above)
Comment (dcharkes) [reply packages/flutter_tools/gradle/src/main/groovy/flutter.groovy:1148]: Ack!

**Pull request merged**